lib/executor/variable_looping_vus.go

```diff
@@ -636,40 +633,88 @@ func (vlv VariableLoopingVUs) Run(ctx context.Context, out chan<- stats.SampleCo
 		currentMaxAllowedVUs = newMaxAllowedVUs
 	}
 
-	handleAllRawSteps := func() bool {
-		for {
+	// This is the step that is the actual gracefulStop at the end
+	lastStep := gracefulExecutionSteps[len(gracefulExecutionSteps)-1]
+	// TODO maybe instead of getting only the last step (above) mergeSteps should return the
+	// remaining gracefulExecutionSteps? or maybe we should refactor this even more ...
+	mergedSteps := mergeSteps(rawExecutionSteps, gracefulExecutionSteps[:len(gracefulExecutionSteps)-1])
+	timer := time.NewTimer(time.Hour * 24)
+	for _, step := range mergedSteps {
+		offsetDiff := step.step.TimeOffset - time.Since(startTime)
+		if offsetDiff > 0 { // wait until time of event arrives // TODO have a mininum
+			timer.Reset(offsetDiff)
 			select {
-			case step, ok := <-rawStepEvents:
-				if !ok {
-					return true
-				}
-				handleNewScheduledVUs(step.PlannedVUs)
-			case step := <-gracefulLimitEvents:
-				if step.PlannedVUs > currentMaxAllowedVUs {
-					// Handle the case where a value is read from the
-					// gracefulLimitEvents channel before rawStepEvents
-					handleNewScheduledVUs(step.PlannedVUs)
-				}
-				handleNewMaxAllowedVUs(step.PlannedVUs)
-			case <-ctx.Done():
-				return false
+			case <-maxDurationCtx.Done():
+				return // exit if context is cancelled
+			case <-timer.C:
+				// now we do a step
 			}
 		}
+		if step.raw {
+			handleNewScheduledVUs(step.step.PlannedVUs)
+		} else {
+			handleNewMaxAllowedVUs(step.step.PlannedVUs)
+		}
 	}
 
-	if handleAllRawSteps() {
-		// Handle any remaining graceful stops
-		go func() {
-			for {
-				select {
-				case step := <-gracefulLimitEvents:
-					handleNewMaxAllowedVUs(step.PlannedVUs)
-				case <-maxDurationCtx.Done():
-					return
-				}
+	go func() {
+		offsetDiff := lastStep.TimeOffset - time.Since(startTime)
+		if offsetDiff > 0 { // wait until time of event arrives // TODO have a mininum
+			timer.Reset(offsetDiff)
+			select {
+			case <-maxDurationCtx.Done():
+				return // exit if context is cancelled
+			case <-timer.C:
+				// now we do a step
 			}
-		}()
-	}
+		}
+		handleNewMaxAllowedVUs(lastStep.PlannedVUs)
+	}()
 
 	return nil
 }
+
+type mergedStep struct {
+	raw  bool              // marks if this is a raw step or gracefulStop one
+	step lib.ExecutionStep // the actual step
+}
+
+// merges raw and gracefulStop steps, giving priority to raw ones if time offset is the same
+// TODO ... thinking about it gracefulExecutionSteps in a lot of cases will lead to nothing so maybe
+// we should skip them here?
+// TODO better name ... zip is not correct, and combine seems even worse ... merge is not great as
+// well though
```

Comment:
Yeah, I just noticed this and another potential optimization we can do with `HasWork()`. Not sure if we should tackle this now, or if I shouldn't just make a new `performance` issue about them...

Reply:
I think that is possible future refactor ;) I really want to get this done with so we can know it's not flaky or buggy, while - yes I would argue that this can be done much earlier, possibly with even better names, and even less code :D

Reply:
:+1: I'll create an issue

Reply:
OK, sounds good.

Reply:
https://github.com/loadimpact/k6/issues/1499 please comment here